[Doc] Add Keycloak login and config docs (#93)
nebari-dev/nebari-docs: docs/docs/how-tos/configure-keycloak-howto.md

---
id: configuring-keycloak
title: Configure Keycloak
description: |
  How to configure Keycloak, add new users, add new groups and update roles.
---

## Introduction to Keycloak

[Keycloak](https://www.keycloak.org/) is the open-source identity and access management tool that comes standard with Nebari. It is used as a centralized location to add new users, create user groups and update roles and permissions. Keycloak can be configured to work with your existing identity provider services such as GitHub, Auth0 and many others.

:::note

> Although Nebari comes hooks to connect with GitHub and Auth0 from the `nebari-config.yaml`, any identity provider that Keycloak can connect to will work. See these [Keycloak docs](https://www.keycloak.org/docs/latest/server_admin/#_identity_broker) for more information.

Reviewer: ```suggestion
Although Nebari allows you to connect with GitHub and Auth0 from the `nebari-config.yaml`, any identity provider that Keycloak can connect to will work. See these [Keycloak docs](https://www.keycloak.org/docs/latest/server_admin/#_identity_broker) for more information.
```

I think this is what we're trying to convey, but please update as you feel is accurate!

:::

In Keycloak, as with many admin tools, you start with an initial login user (`root`) that has the ability to administer and create new users. The `root` user is a Keycloak-specific user. It can only be used to login and manage the Keycloak identity management section of Nebari.

:::warning

> Trying to login to JupyterHub using this `root` user will fail. Currently this is the only user specific to the Keycloak admininstration console, therefore anyone wanting to add users, new groups, etc, will need to use this user and login here: `https://{your-nebari-domain}/auth/admin/`.

Reviewer: ```suggestion
Trying to login to JupyterHub using this `root` user will fail. Currently, this is the only user with access to the Keycloak administration console, therefore anyone wanting to add users, new groups, etc., will need to use this user and login here: `https://{your-nebari-domain}/auth/admin/`.
```

4. In the Password section, click the **Update** button. This will guide you through entering your existing root password, and then creating a new password.

   ![Keycloak root user page - account security, update password](/img/how-tos/keycloak_root_user_update_password.png)

:::warning
The `security.keycloak.initial_root_password` field in `nebari-config.yaml` has no effect after changing the `root` password. If you redeploy Nebari it **will not reset** the password back to the old one (or anything else that might be in the field in your YAML file). We strongly recommend you delete this field to prevent later confusion.
:::

## Adding a Nebari user

No Nebari user is created for you during deployment, you will need to add a Nebari user manually in order to [login](/how-tos/login-keycloak#how-to-authenticate-using-keycloak-and-launch-your-server) to JupyterHub and access the other Nebari services.

If you have chosen to use GitHub, Auth0 or any other single-sign-on, you must ensure the value you enter in Keycloak under `Username` exactly matches the username from GitHub or Auth0, respectively.

### Add user using Keycloak admin console

> To add a Nebari user from the Keycloak admin console, visit https://{your-nebari-domain}/auth/admin/ and login using the username `root`, as shown above.

Reviewer: ```suggestion
To add a Nebari user from the Keycloak admin console, visit `https://{your-nebari-domain}/auth/admin/` and login using the username `root`, as shown above.
```

:::

The `nebari` realm is selected by default, we strongly recommend leaving it as is.

Steps to create a new user:

1. Click **Users** along the left-hand side of the page.

2. Click the **Add user** button and you will see the new user form:

   ![Keycloak add user tab screenshot - new user form](/img/how-tos/keycloak_add_users.png)

3. Fill out the three fields outlined above. These are **Username**, **Email**, and **Groups**. (See explanation below). Then click **save**.

   - Username: Depending on the authentication provider selected ('password', 'GitHub' or 'Auth0'), the values entered into the **Username** field will differ slightly. The following table outlines
     those differences:

     |          | Password          | GitHub            | Auth0                 |
     | -------- | ----------------- | ----------------- | --------------------- |
     | Username | _unique username_ | _GitHub username_ | _Email to login with_ |

> - Once the **Username** field is updated, add a valid email address in the **Email** field.

Reviewer: ```suggestion
   - Once the **Username** field is updated, add a valid email address in the **Email** field.
```

  :::note
  Although not required, users may not be able to log in to Grafana if this field isn't properly set.
  :::

> - Associate the user with one or more of the **Groups**. Out of the box, Nebari is deployed with the following groups: `admin`, `developer` and `analyst` (see the [Roles x Groups](#roles-x-groups) section below for more details).

Reviewer: ```suggestion
   - Associate the user with one or more of the **Groups**. Out of the box, Nebari is deployed with the following groups: `admin`, `developer` and `analyst` (see the [Roles x Groups](#roles-x-groups) section below for more details).
```

> 1. Click **Save**.

Reviewer: ```suggestion
4. Click **Save**.
```

:::note
If using Auth0, Github or any other identity provider, the password field is not required.
:::

![Keycloak add user > credentials tab screenshot - set password](/img/how-tos/keycloak_user_password.png)

### Add user from the command line

:::warning
If you [changed the initial root password for Keycloak](#change-keycloak-root-password) this method will not work.
:::

To make adding users easier for new Nebari deployments, we've created a CLI command to help you.

```shell
nebari keycloak -c nebari-config.yaml adduser <username> <password>
```

This will create a new user `<username>` under the `analyst` group, with the initial password provided. Omit the password completely if you are using GitHub or Auth0. It will also add a dummy email (i.e. `username@your-domain`) to the **Email** field.

## Login to Nebari

Your new user can now log in to Nebari, visit your provided Nebari domain URI which will take you to the login form page and follow the [How to login in to Nebari and start a server](/how-tos/login-keycloak).

![Nebari - Log in to Keycloak page](/img/how-tos/nebari_login_screen.png)

## In-depth look at Roles and Groups

> Groups represent a collection of users that perform similar actions and therefore require similar permissions. By default, Nebari is deployed with the following groups: `admin`, `developer`, and `analyst` (in roughly descending order of permissions and scope).

Reviewer: This "in-depth look" is a great section to have, thank you for writing it!

Can we also include a description of "Roles" here similar to what we have for "Groups"? -- I think I don't fully understand how the Realm role and client roles are connected, so maybe a sentence about that will also help. 

(This can be a part of future work as a part of our Glossary of terms as well)

Reviewer 2: Definitely! I added a small paragraph trying to outline the differences but let me know if this is unclear 👍

Reviewer: Thank you!

:::caution
The role `jupyterhub_admin` gives users special permissions to JupyterHub and should be applied judiciously. As mentioned in the table above, a JupyterHub admin is able to impersonate other users and view the contents of their home folder. For more details, read through the [JupyterHub documentation](https://z2jh.jupyter.org/en/stable/jupyterhub/customizing/user-management.html#admin-users).
:::

To create new groups or modify (or delete) existing groups, log in as `root` and click **Groups** on the left-hand side.

> As an example, we create a new group named `conda-store-manager`. This group will have administrator access to the [Conda-Store service].

Reviewer: ```suggestion
As an example, we create a new group named `conda-store-manager`. This group will have administrator access to the [Conda-Store service](https://conda-store.readthedocs.io/en/latest/).
```
